File: pc/McController.Server/Input/JoystickToWasdMapper.cs
using McController.Server.Config;

namespace McController.Server.Input;

/// <summary>
/// Translates a normalized joystick position [-1, 1] into WASD key state.
///
/// Convention: y > 0 means forward (W). Android flips screen-Y before sending.
///
/// Hysteresis (enter/exit thresholds) prevents jitter when the stick rests near
/// the activation boundary. If the user reverses direction across zero, the
/// previously-held key is released first.
/// </summary>
public sealed class JoystickToWasdMapper
{
    private readonly IInputInjector _injector;
    private readonly ServerConfig _config;
    private readonly object _lock = new();

    private bool _w, _a, _s, _d;

    public JoystickToWasdMapper(IInputInjector injector, ServerConfig config)
    {
        _injector = injector;
        _config = config;
    }

    public void Update(float x, float y)
    {
        lock (_lock)
        {
            UpdateAxis(y, Scancodes.W, Scancodes.S, ref _w, ref _s);
            UpdateAxis(x, Scancodes.D, Scancodes.A, ref _d, ref _a);
        }
    }

    public void ReleaseAll()
    {
        lock (_lock)
        {
            if (_w) { _injector.Key(Scancodes.W, false); _w = false; }
            if (_a) { _injector.Key(Scancodes.A, false); _a = false; }
            if (_s) { _injector.Key(Scancodes.S, false); _s = false; }
            if (_d) { _injector.Key(Scancodes.D, false); _d = false; }
        }
    }

    private void UpdateAxis(
        float v, ushort posKey, ushort negKey,
        ref bool posDown, ref bool negDown)
    {
        var cfg = _config.Movement;
        var abs = Math.Abs(v);

        // Use <= so that abs=0 (release event) always triggers release,
        // even when the user has set DeadZone or ExitThreshold to 0.
        if (abs <= cfg.DeadZone)
        {
            if (posDown) { _injector.Key(posKey, false); posDown = false; }
            if (negDown) { _injector.Key(negKey, false); negDown = false; }
            return;
        }

        if (v > 0)
        {
            // Target = posKey. Release opposite if held.
            if (negDown) { _injector.Key(negKey, false); negDown = false; }
            if (!posDown && abs > cfg.EnterThreshold)
            {
                _injector.Key(posKey, true);
                posDown = true;
            }
            else if (posDown && abs <= cfg.ExitThreshold)
            {
                _injector.Key(posKey, false);
                posDown = false;
            }
        }
        else
        {
            // Target = negKey.
            if (posDown) { _injector.Key(posKey, false); posDown = false; }
            if (!negDown && abs > cfg.EnterThreshold)
            {
                _injector.Key(negKey, true);
                negDown = true;
            }
            else if (negDown && abs <= cfg.ExitThreshold)
            {
                _injector.Key(negKey, false);
                negDown = false;
            }
        }
    }
}

File: mac/McController/Core/Input/JoystickToWasdMapper.swift
import Foundation

/// Translates a normalized joystick position [-1, 1] into WASD key state.
/// Convention: y > 0 means forward (W) — the Android client flips
/// screen-Y before sending. Mirrors `JoystickToWasdMapper.cs`.
///
/// Hysteresis (enter/exit thresholds) prevents jitter when the stick
/// rests near the activation boundary. Crossing zero releases the
/// previously-held opposite key first.
///
/// The `<=` comparison in dead-zone / exit checks is intentional: with
/// every threshold at 0, a release event (`abs == 0`) must still lift
/// the key, otherwise a "stuck-A" regression appears. There's a parity
/// test for this on the Windows side (see
/// `JoystickToWasdMapperTests.AllThresholdsZero_ReleaseAtZero_StillReleases`).
final class JoystickToWasdMapper {

    private let injector: InputInjector
    private let config: ServerConfig
    private let lock = NSLock()

    private var wDown = false
    private var aDown = false
    private var sDown = false
    private var dDown = false

    init(injector: InputInjector, config: ServerConfig) {
        self.injector = injector
        self.config = config
    }

    func update(x: Float, y: Float) {
        lock.lock(); defer { lock.unlock() }
        updateAxis(value: y, posKey: forwardKey(), negKey: backKey(),
                   posDown: &wDown, negDown: &sDown)
        updateAxis(value: x, posKey: rightKey(), negKey: leftKey(),
                   posDown: &dDown, negDown: &aDown)
    }

    func releaseAll() {
        lock.lock(); defer { lock.unlock() }
        if wDown { injector.key(forwardKey(), down: false); wDown = false }
        if aDown { injector.key(leftKey(),    down: false); aDown = false }
        if sDown { injector.key(backKey(),    down: false); sDown = false }
        if dDown { injector.key(rightKey(),   down: false); dDown = false }
    }

    // Resolved on every call instead of cached: the Key Bindings page
    // edits `config.movementKeys` in place and the next joystick sample
    // should already honour the new mapping, no profile reload needed.
    // `KeyCodes.resolve` accepts symbolic names ("w", "jump"), Windows
    // hex scancodes ("0x11"), and decimal scancodes. If somehow blank or
    // invalid we fall back to the matching WASD direction.
    private func forwardKey() -> UInt16 {
        KeyCodes.resolve(config.movementKeys.forward) ?? KeyCodes.w
    }
    private func backKey() -> UInt16 {
        KeyCodes.resolve(config.movementKeys.back) ?? KeyCodes.s
    }
    private func leftKey() -> UInt16 {
        KeyCodes.resolve(config.movementKeys.left) ?? KeyCodes.a
    }
    private func rightKey() -> UInt16 {
        KeyCodes.resolve(config.movementKeys.right) ?? KeyCodes.d
    }

    private func updateAxis(value v: Float, posKey: UInt16, negKey: UInt16,
                            posDown: inout Bool, negDown: inout Bool) {
        let cfg = config.movement
        let abs = Swift.abs(v)

        if abs <= cfg.deadZone {
            if posDown { injector.key(posKey, down: false); posDown = false }
            if negDown { injector.key(negKey, down: false); negDown = false }
            return
        }

        if v > 0 {
            if negDown { injector.key(negKey, down: false); negDown = false }
            if !posDown && abs > cfg.enterThreshold {
                injector.key(posKey, down: true); posDown = true
            } else if posDown && abs <= cfg.exitThreshold {
                injector.key(posKey, down: false); posDown = false
            }
        } else {
            if posDown { injector.key(posKey, down: false); posDown = false }
            if !negDown && abs > cfg.enterThreshold {
                injector.key(negKey, down: true); negDown = true
            } else if negDown && abs <= cfg.exitThreshold {
                injector.key(negKey, down: false); negDown = false
            }
        }
    }
}
